application/src/main/java/org/togetherjava/tjbot/commands/code/CodeMessageAutoDetection.java

```diff
@@ -0,0 +1,80 @@
+import net.dv8tion.jda.api.entities.channel.ChannelType;
+import net.dv8tion.jda.api.entities.channel.concrete.ThreadChannel;
+import net.dv8tion.jda.api.events.message.MessageReceivedEvent;
+
+import org.togetherjava.tjbot.commands.MessageReceiverAdapter;
+import org.togetherjava.tjbot.commands.utils.CodeFence;
+import org.togetherjava.tjbot.commands.utils.MessageUtils;
+import org.togetherjava.tjbot.config.Config;
+
+import java.util.Optional;
+import java.util.function.Predicate;
+import java.util.regex.Pattern;
+
+/**
+ * Automatically detects messages that contain code and registers them at {@link CodeMessageHandler}
+ * for further processing.
+ */
+public final class CodeMessageAutoDetection extends MessageReceiverAdapter {
+    private static final long MINIMUM_LINES_OF_CODE = 3;
+
+    private final CodeMessageHandler codeMessageHandler;
+
+    private final Predicate<String> isStagingChannelName;
+    private final Predicate<String> isOverviewChannelName;
+
+    /**
+     * Creates a new instance.
+     * 
+     * @param config to figure out whether a message is from a help thread
+     * @param codeMessageHandler to register detected code messages at for further handling
+     */
+    public CodeMessageAutoDetection(Config config, CodeMessageHandler codeMessageHandler) {
+        super(Pattern.compile(".*"));
+
+        this.codeMessageHandler = codeMessageHandler;
+
+        isStagingChannelName = Pattern.compile(config.getHelpSystem().getStagingChannelPattern())
+            .asMatchPredicate();
+        isOverviewChannelName = Pattern.compile(config.getHelpSystem().getOverviewChannelPattern())
+            .asMatchPredicate();
+    }
+
+    @Override
+    public void onMessageReceived(MessageReceivedEvent event) {
+        if (event.isWebhookMessage() || event.getAuthor().isBot() || !isHelpThread(event)) {
+            return;
+        }
+
+        Message originalMessage = event.getMessage();
+        String content = originalMessage.getContentRaw();
+
+        Optional<CodeFence> maybeCode = MessageUtils.extractCode(content);
+        if (maybeCode.isEmpty()) {
+            // There is no code in the message, ignore it
+            return;
+        }
+
+        long amountOfCodeLines =
+                maybeCode.orElseThrow().code().lines().limit(MINIMUM_LINES_OF_CODE + 1).count();
+        if (amountOfCodeLines < MINIMUM_LINES_OF_CODE) {
+            return;
+        }
+
+        codeMessageHandler.addAndHandleCodeMessage(event.getMessage());
+    }
+
+    private boolean isHelpThread(MessageReceivedEvent event) {
+        if (event.getChannelType() != ChannelType.GUILD_PUBLIC_THREAD) {
+            return false;
+        }
+
+        ThreadChannel thread = event.getChannel().asThreadChannel();
+        String rootChannelName = thread.getParentChannel().getName();
+        return isStagingChannelName.test(rootChannelName)
+                || isOverviewChannelName.test(rootChannelName);
+    }
```

Comment:
This method is on a multitude of places. We might want to add a HelpUtil, doesn't have to be in this PR but feel free to

Reply:
yeah saw that as well. its duplicated 3 times...

Reply:
its a bit difficult to properly extract it, since the method is primarly based on two config patterns.

so either it gets them via arg, but then barely provides any real utility (it essentially just tests the two patterns and thats it), or the utility is not static (so that it can read and hold the config itself).

id probably vote for moving this to a different PR, i dont have a good solution right out of the pocket for today

Reply:
Sure